cli/cli.go

```diff
@@ -37,14 +38,49 @@ func printVersion(c *cli.Context) {
 }
 
 func action(c *cli.Context) {
-	if os.Getenv(plugins.PluginInputPluginModeKey) == string(plugins.TriggerMode) {
-		sendAnalytics()
+	if os.Getenv(plugins.PluginInputPluginModeKey) != string(plugins.TriggerMode) {
+		log.Errorf("Required envs not set: only Bitrise CLI is intended to send build run analytics")
+
+		if err := cli.ShowAppHelp(c); err != nil {
+			failf("Failed to show help, error: %s", err)
+		}
+		return
+	}
+
+	if enabled, err := isAnalyticsEnabled(); err != nil {
+		failf("Failed to check if analytics enabled: %s", err.Error())
+	} else if !enabled {
+		log.Debugf("Build run analytics disabled, terminating...")
+		return
+	}
+
+	if warn, err := checkFormatVersion(os.Getenv(plugins.PluginInputFormatVersionKey), models.Version); err != nil {
+		failf(err.Error())
+	} else if len(warn) > 0 {
+		log.Warnf(warn)
+	}
+
+	var t SourceType
+	if available, err := hasContent(os.Stdin); err != nil {
+		failf("Failed to check if analytics enabled: %s", err.Error())
+	} else if available {
+		log.Debugf("stdin payload provided")
+		t = StdinSource
+	} else if os.Getenv(plugins.PluginInputPayloadKey) != "" {
+		log.Debugf("env payload provided")
+		t = EnvSource
+	} else {
+		log.Errorf("No stdin data nor env data provided: only Bitrise CLI is intended to send build run analytics")
+
+		if err := cli.ShowAppHelp(c); err != nil {
+			failf("Failed to show help, error: %s", err)
+		}
 		return
 	}
 
-	if err := cli.ShowAppHelp(c); err != nil {
-		log.Errorf("Failed to show help, error: %s", err)
-		os.Exit(1)
+	source := PayloadSourceFactory(t)
```

Comment:
NOTE: makes sense, however still I feel the lack of control because you are using a global variable(s) separately in your pre-cosntruct phase and in you run-phase

Reply:
at some point in the code, I need to write os.Stdin as the plugin needs to run the standard input, where would it be a better place for it?